File: asaf-retrofit/src/main/java/co/early/asaf/retrofit/ErrorHandler.java
package co.early.asaf.retrofit;

import okhttp3.Request;
import retrofit2.Response;

/**
 *
 * @param <M> Error message class to be used throughout the app, such as an enum
 */
public interface ErrorHandler<M> {
    /**
     *
     * @param t throwable that caused the error
     * @param errorResponse error response from the server, the body of which may represent a custom error
     * @param customErrorClazz custom error class expected from the errorResponse
     * @param <CE> class type of the custom error
     * @return the parsed error from the server
     */
    <CE extends MessageProvider<M>> M handleError(Throwable t, Response errorResponse, Class<CE> customErrorClazz, Request originalRequest);
}

File: fore-kt/fore-kt-network/src/main/java/co/early/fore/kt/net/ktor/ErrorHandler.kt
package co.early.fore.kt.net.ktor

import co.early.fore.net.MessageProvider

/**
 *
 * @param <F> Failure message class to be used throughout the app (or the data layer in the case of
 * clean architecture) such as an enum or a data class
 */
interface ErrorHandler<F> {
    /**
     *
     * @param t throwable that caused the error
     * @param customErrorClazz custom error class expected from the errorResponse, may be null
     * @param <CE> class type of the custom error if specified
     * @return the parsed error from the server
    </CE> */
    suspend fun <CE : MessageProvider<F>> handleError(t: Throwable, customErrorKlazz: kotlin.reflect.KClass<CE>?): F
}
